File: Project 2/projprojectile.py
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  5 23:30:25 2020

@author: danie
"""

# air resistance? in physics? unheard of!

# import packages
from scipy.integrate import odeint as ode
import numpy as np
import math as ma
import matplotlib.pyplot as pp

# define constants
g = 9.81;
rho = 1.225; # air density from Google
C = .1; # drag coeffecient of a sphere from Google
A = 5; # randomly decided number
m = 5; # randomly decided number

# calculate drag force coeffecient for use in motion equations
fdrag = (C*rho*A)/(2*m);

# write function
def projprojectile(hi=10,vi=60,angi=45):
    # accepts inputs of initial height in m, initial velocity in m/s, and
    # initial angle in deg and returns max height, range, time of flight, and
    # a graph of trajectory
    # error messages for inavalid inputs
    if hi < 0:
        print('height must be 0 or greater')
    if vi <= 0:
        print('velocity must be greater than 0')
    if angi <=0 or angi >= 90:
        print('angle must be between 0 and 90 degrees')
    # convert deg to rad
    angrad = np.radians(angi);
    # calculate components of initial velocity
    vix = vi*(ma.cos(angrad));
    viy = vi*(ma.sin(angrad));
    # create arrays
    x = 0;
    y = hi;
    xarray = np.array([x, vix]);
    yarray = np.array([y, viy]);
    # calculate time and timespan
    t = (np.sqrt(viy**2+2*g*hi)+viy)/g;
    timespan = np.linspace(0,t);
    # make functions that return arrays
    # thanks to Ian's code for help with this
    def odex(x,t):
        vx = x[1];
        return np.array([vx,-fdrag*vx**2]);
    def odey(y,t):
        vy = y[1];
        return np.array([vy,-g-fdrag*vy**2]);
    # solve odes
    xsol = ode(odex,xarray,timespan);
    ysol = ode(odey,yarray,timespan);
    # calculate positions
    xpos = xsol[:,0];
    ypos = ysol[:,0];
    # calculate output values
    hmax = np.amax(ypos);
    range = xpos[len(ypos)-1];
    time = timespan[len(ypos)-1];
    # graph
    traj = pp.plot(x[:,0],y[:,0]);
    pp.xlabel('range (m)');
    pp.ylabel('height (m)');
    pp.title('trajectory of a projectile with drag');
    pp.xlim(0,range+1);
    pp.ylim(0,hmax+1);
    return hmax,range,time,traj

File: Project 2/projprojectile.m
function[hmax,range,time,traj] = projprojectile(hi,vi,angi)
% accepts inputs of initial height in m, initial velocity in m/s, and
% initial angle in deg and returns max height, range, time of flight, and a
% graph of trajectory

% define constants
g = 9.81;
rho = 2.225; % air density from Google
C = .1; % drag coeffecient of a sphere from Google
A = 5; % randomly decided number
m = 5; % randomly decided number

% calculate drag force coeffecient for use in motion equations
fdrag = (C*rho*A)/(2*m);

% set default values
if nargin == 0
    hi = 10;
    vi = 60;
    angi = 45;
end
    
% error messages for invalid inputs
if hi < 0
    error('height must be 0 or greater')
end
if vi <= 0
    error('velocity must be greater than 0')
end
if angi <= 0 || angi >= 90
    error('angle must be between 0 and 90 degrees')
end

% calculate components of initial velocity
vix = vi*cosd(angi);
viy = vi*sind(angi);

% create arrays
x = 0;
y = hi;
xarray = [x vix];
yarray = [y viy];

% calculate time and timespan
t = (sqrt(viy^2+2*g*hi)+viy)/g;
timespan = linspace(0, t);

% solve odes
options = odeset('refine', 10);
[xt, xsol] = ode45(@odex,timespan,vix,options,fdrag);
[yt, ysol] = ode45(@odey,timespan,viy,options,fdrag,g);

% make functions that return arrays
    function[dx] = odex(xt,xsol,fdrag)
        dx = [x(2);-fdrag*xsol(2)^2];
    end
    function[dy] = odey(yt,ysol,fdrag,g)
        dy = [y(2);-g-fdrag*ysol(2)^2];
    end

% calculate outputs
ypos = y(y(:,1)>=0);
numypos = numel(ypos);
hmax = max(ypos);
range = x(ynum);
time = yt(numypos);

% graph
xspan = x(1:numypos);
yspan = y(1:numypos);
traj = plot(xspan,yspan);
xlabel('range (m)');
ylabel('height (m)');
title('trajectory of a projectile with drag');

end